### PR TITLE
improve gamepad stash control, fix #7019

### LINKED ISSUE #7019: [Feature Request]: Stash objects “center of gravity”
### Feature Type

Quality of Life

### Describe

I’m not quite sure if this is a bug or just a missing feature: when you move the gamepad cursor in the inventory over objects bigger than one square, it jumps straight in the middle of them, so exactly in the center of a book (4 squares) or an armor (6 squares). This is very convenient and useful to save time.
It doesn’t happen in the stash, where the cursor keeps moving by a square every time, even on bigger object. This requires several joypad movements to pass from a big weapon to the next one.
Is it possible to have the same inventory behavior everywhere?

### PR DESCRIPTION
**Pull request by Sophie452**
Should improve gamepad control in stash, see #7019
First time contribution, so be gentle :)

### CODE REVIEW

File: Source/controls/plrctrls.cpp [hunk 1]
```diff
@@ -1171,6 +1185,17 @@ void StashMove(AxisDirection dir)
...
 		// Stash coordinates are all the top left of the cell, so we need to shift the mouse to the center of the held item
 		// or the center of the cell if we have a hand cursor (itemSize will be 1x1 here so we can use the same calculation)
+		if (holdItem.isEmpty()) {
+			const StashStruct::StashCell itemIdAtActiveStashSlot = Stash.GetItemIdAtPosition(ActiveStashSlot);
+			if (itemIdAtActiveStashSlot != StashStruct::EmptyCell) {
+				const Item stashItem = Stash.stashList[itemIdAtActiveStashSlot];
+				const Point firstSlotOnItem = FindFirstStashSlotOnItem(itemIdAtActiveStashSlot);
+				itemSize = GetInventorySize(stashItem);
+				mousePos = GetStashSlotCoord(firstSlotOnItem);
+				ActiveStashSlot = FindClosestStashSlot(mousePos);
```
Comment (ephphatha) [inline Source/controls/plrctrls.cpp:1195]: This line is actually unnecessary, it causes a behaviour difference compared to moving through the inventory. If you comment it out the gamepad navigation "remembers" the last column/row that was active when moving along staggered items. As-is you end up laddering to the top left cell of the stash.

https://github.com/user-attachments/assets/4888adea-d590-414b-aab9-134e1ec24ecc



https://github.com/user-attachments/assets/ceef7785-8e35-474b-b8ba-3086e1aad768

**Review by ephphatha — CHANGES_REQUESTED** [review]
Nice to see some progress on that issue :D (especially since I've been too lazy to do anything like I said I would ages back...). The behaviour is a definite improvement compared to what we currently have.

As part of https://github.com/diasurgical/devilutionX/pull/6536 I added some code in PerformPrimaryAction (see https://github.com/Sophie452/devilutionX/blob/c641b49d73a547fcf953ed305936c9414d525f45/Source/controls/plrctrls.cpp#L1897-L1904) that shifts the cursor to the bottom right cell after pasting an item. That's no longer necessary with this PR, we want the cursor to stay in the centre of the item.

**Comment by AJenbo** [comment]
@Sophie452 nice work, if you have time to address the comments from @ephphatha I think this will be good to merge. You can ignore the failing PS4 build, it's a know issue with a required package.

File: Source/controls/plrctrls.cpp [hunk 2]
```diff
@@ -1135,11 +1135,18 @@ void StashMove(AxisDirection dir)
 	}
 
 	if (dir.x == AxisDirectionX_LEFT) {
-		if (ActiveStashSlot.x > 0)
-			ActiveStashSlot.x--;
+		if (ActiveStashSlot.x > 0) {
+			const StashStruct::StashCell itemIdAtActiveStashSlot = Stash.GetItemIdAtPosition(ActiveStashSlot);
+			do {
+				ActiveStashSlot.x--;
+			} while (ActiveStashSlot.x > 0 && itemIdAtActiveStashSlot == Stash.GetItemIdAtPosition(ActiveStashSlot) && itemIdAtActiveStashSlot != StashStruct::EmptyCell && holdItem.isEmpty());
```
[the remark this answers is not in the document]
Comment (Sophie452) [reply Source/controls/plrctrls.cpp]: Better? :)

File: Source/controls/plrctrls.cpp [hunk 3]
```diff
@@ -1171,6 +1185,17 @@ void StashMove(AxisDirection dir)
...
+				const Item stashItem = Stash.stashList[itemIdAtActiveStashSlot];
+				const Point firstSlotOnItem = FindFirstStashSlotOnItem(itemIdAtActiveStashSlot);
+				itemSize = GetInventorySize(stashItem);
+				mousePos = GetStashSlotCoord(firstSlotOnItem);
+				ActiveStashSlot = FindClosestStashSlot(mousePos);
+			}
+		}
+
 		mousePos += Displacement { itemSize.width * INV_SLOT_HALF_SIZE_PX, itemSize.height * INV_SLOT_HALF_SIZE_PX };
 		SetCursorPos(mousePos);
```
[the remark this answers is not in the document]
Comment (Sophie452) [reply Source/controls/plrctrls.cpp]: Tried to copy relevant comment lines

File: Source/controls/plrctrls.cpp (hunk 1 above)
Comment (Sophie452) [reply Source/controls/plrctrls.cpp]: removed

**Comment by Sophie452** [comment]
Thanks for your thorough review. I tried to corrrect all points.

**Comment by AJenbo** [comment]
Also please run clag-format to make sure the changes conform to the style guide

**Comment by Sophie452** [comment]
Sorry that you had to wait. The improvements have been completed and added the clang-format. Thanks for your help :)

**Comment by AJenbo** [comment]
Thanks to both of you

**Pull request merged**